File: packages/shared/ts/src/events.ts
// SSE 事件协议（规格书 §5「SSE 协议」）
// 变更时必须同步 packages/shared/python/comfyportal_shared/events.py

import type { TaskStatus } from "./task-state";

/** SSE 的 event: 字段值，前端按此分发 */
export type SSEEventType = "status" | "progress" | "done" | "error";

export interface ArtifactSummary {
  id: number;
  kind: string;
  url: string;
  width: number;
  height: number;
}

/** event: status */
export interface StatusPayload {
  state: TaskStatus;
  position?: number; // 队列位置，queued 时存在
  comfy_prompt_id?: string; // running 时存在
}

/** event: progress */
export interface ProgressPayload {
  pct: number; // 0-100
  node: string; // 当前节点，如 "KSampler"
  step: number;
  max_steps: number;
}

/** event: done */
export interface DonePayload {
  state: "success";
  artifacts: ArtifactSummary[];
}

/** event: error */
export interface ErrorPayload {
  state: "failed";
  error: string;
}

export type SSEPayload = StatusPayload | ProgressPayload | DonePayload | ErrorPayload;

File: packages/shared/python/comfyportal_shared/events.py
"""SSE 事件 payload（规格书 §5「SSE 协议」）。

与 TS 侧 packages/shared/ts/src/events.ts 保持同步。
event: 字段值为 status/progress/done/error，对应下方各 Payload。
"""

from __future__ import annotations

from typing import Literal, Optional

from pydantic import BaseModel

from .task_state import TaskStatus


class ArtifactSummary(BaseModel):
    id: int
    kind: str
    url: str
    width: int
    height: int


class StatusPayload(BaseModel):
    state: TaskStatus
    position: Optional[int] = None  # queued 时存在
    comfy_prompt_id: Optional[str] = None  # running 时存在


class ProgressPayload(BaseModel):
    pct: float  # 0-100
    node: str  # 当前节点，如 "KSampler"
    step: int
    max_steps: int


class DonePayload(BaseModel):
    state: Literal["success"]
    artifacts: list[ArtifactSummary]


class ErrorPayload(BaseModel):
    state: Literal["failed"]
    error: str
